### PR TITLE
Update optimizers to work with newer interface.

### PR DESCRIPTION
**Pull request by mstechly**

### CODE REVIEW

**Comment by simonwa7** [comment]
I will wait to review until https://github.com/zapatacomputing/z-quantum-core/pull/351 is settled.

File: src/python/zquantum/optimizers/cma_es_optimizer.py [hunk 1]
```diff
@@ -10,58 +8,49 @@
 )
 from scipy.optimize import OptimizeResult
 import cma
+from typing import Dict, Optional
 
 
 class CMAESOptimizer(Optimizer):
-    def __init__(self, options):
+    def __init__(
+        self, sigma_0: float, options: Optional[Dict] = None, recorder=_recorder
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/cma_es_optimizer.py:16]: Typing needed for recorder.

File: src/python/zquantum/optimizers/grid_search/_grid_search.py [hunk 2]
```diff
@@ -12,48 +12,37 @@
 
 
 class GridSearchOptimizer(Optimizer):
-    def __init__(self, grid: ParameterGrid, options: Optional[Dict] = None):
+    def __init__(self, grid: ParameterGrid, recorder=_recorder):
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/grid_search/_grid_search.py:15]: Typing needed for recorder.

File: src/python/zquantum/optimizers/scipy_optimizer.py [hunk 3]
```diff
@@ -1,63 +1,55 @@
...
 
 
 class ScipyOptimizer(Optimizer):
     def __init__(
         self,
         method: str,
         constraints: Optional[Tuple[Dict[str, Callable]]] = None,
-        options=None,
+        options: Optional[Dict] = None,
+        recorder=_recorder,
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/scipy_optimizer.py:19]: Typing needed for recorder.

File: src/python/zquantum/optimizers/basin_hopping.py [hunk 4]
```diff
@@ -49,30 +49,31 @@ def __init__(
...
         if hasattr(cost_function, "gradient") and callable(
             getattr(cost_function, "gradient")
         ):
             jacobian = cost_function.gradient
-        if self.minimizer_kwargs is not None:
-            if self.minimizer_kwargs.get("options", None) is not None:
-                self.minimizer_kwargs["options"]["jacobian"] = jacobian
+        if (
+            self.minimizer_kwargs is not None
+            and self.minimizer_kwargs.get("options", None) is not None
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/optimizers/basin_hopping.py:74]: I would leave it as it is right now, as we might change the logic for gradients soon anyway.

File: src/python/zquantum/optimizers/basin_hopping.py [hunk 5]
```diff
@@ -21,13 +21,13 @@ def __init__(
         interval: int = 50,
         disp: bool = False,
         niter_success: Union[int, None] = None,
+        recorder=_recorder,
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/optimizers/basin_hopping.py]: Fixed in https://github.com/zapatacomputing/z-quantum-optimizers/pull/58/commits/d4fba8b8bd7930fd68bc5a0388d153201c79bc4b

File: src/python/zquantum/optimizers/basin_hopping.py [hunk 6]
```diff
@@ -1,17 +1,17 @@
+import numpy as np
 from zquantum.core.interfaces.optimizer import (
     Optimizer,
     optimization_result,
     construct_history_info,
 )
-from zquantum.core.history.recorder import recorder
-from typing import Callable, Dict, Union
+from zquantum.core.history.recorder import recorder as _recorder, CallableWithGradient
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/optimizers/basin_hopping.py]: Fixed in https://github.com/zapatacomputing/z-quantum-optimizers/pull/58/commits/d4fba8b8bd7930fd68bc5a0388d153201c79bc4b

File: src/python/zquantum/optimizers/cma_es_optimizer.py (hunk 1 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/cma_es_optimizer.py]: Fixed in https://github.com/zapatacomputing/z-quantum-optimizers/pull/58/commits/d4fba8b8bd7930fd68bc5a0388d153201c79bc4b

File: src/python/zquantum/optimizers/grid_search/_grid_search.py (hunk 2 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/grid_search/_grid_search.py]: Fixed in https://github.com/zapatacomputing/z-quantum-optimizers/pull/58/commits/d4fba8b8bd7930fd68bc5a0388d153201c79bc4b

**Review by simonwa7 — APPROVED** [review]

**Pull request merged**